File: C/credit.c
#include <stdio.h>
#include <cs50.h>

void verify(long c, long len1);

int main(void)
{
    //variables
    long card = get_long("What's the card number? ");
    long duble = 0;
    long n1 = 0;
    long n2 = 0;
    long n3 = 0;
    long num = 0;
    long len = 0;
    long digit = 0;
    long digit2 = 0;
    long digit3 = 0;
    
    //separating card numbers-----------------
    for (; card > 0;)
    {
        //calculate len-------------
        len++;
        digit = card % 10;
        card = (card - digit) / 10;
        
        //math operations---------------
        if (duble == 1)
        {
            n1 = 2 * digit;
            digit2 = n1 % 10;
            n1 = ((n1 - digit2) / 10) + digit2;
            n3 += n1;

            duble--;
        }
        //math operations---------------
        else
        {
            n2 += digit;
            duble++;
        }
        
        //Agency----------------
        if (card < 100 && card > 36)
        {
            num = card;
            if (num < 50 && num > 39)
            {
                num /= 10;
            }
        }
    }
    
    //verifying math operations-----------------
    n3 += n2;
    digit3 = n3 % 10;
    
    if (digit3 == 0)
    {
        verify(num, len);
    }
    else
    {
        printf("INVALID\n");
    }
}

//verifying finally-----------------------------
void verify(long c, long len1)
{
    //Verifying the len and agency number-----------
    if (c == 4 && (len1 == 13 || len1 == 16))
    {
        printf("VISA\n");
    }
    else if (c > 50 && c < 56 && len1 == 16)
    {
        printf("MASTERCARD\n");
    }
    else if (len1 == 15 && (c == 34 || c == 37))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Python/credit.py
from cs50 import get_string

card = 0
lista = []
le = 0
check = False
# Vendo se o que foi fornecido e maior que 0
while not check:
    card = str(input("Number: "))
    
    # Criando uma lista
    for num in card:
        lista.append(int(num))
    le = len(lista)
    if le > 0:
        check = True
    else:
        check = False

sub = 0
lista2 = []
lista3 = []
lista4 = []
soma = 0
soma2 = 0
s: str = ""

# Pegando os numeros apartir do ultimo
for num in range(1, le+1, 2):
    sub = le - num - 1
    
    if sub < 0:
        break
    
    lista2.append(lista[sub])

# Adicionado os numeros multiplicados por dois em outra lista
for num in lista2:
    lista3.append(num * 2)

# Somando todos os digitos da lista2
for num in lista3:
    s = str(num)
    for num1 in s:
        soma += int(num1)

# Pegando os numeos que nao foram multiplicados por 2
for num in range(1, le+1, 2):
    lista4.append(lista[le - num])

lista4.append(soma)
soma2 = sum(lista4)

# Limpando algumas listas para utilizalas de novo

lista3.clear()
lista2.clear()
lista3.append(soma2)

# Separando o resultado

for num in lista3:
    s = str(num)
    for num1 in s:
        lista2.append(num1)

lista2.sort()

# Juntando os digitos para verificar

valor1 = lista[0]
valor2 = lista[1]
valor = (valor1 * 10) + valor2

# Vai verificar de qual empresa e
if lista2[0] == "0":
    if le == 15:
        if valor == 34 or valor == 37:
            print("AMEX")
        else:
            print("INVALID")
    
    elif le == 16:
        if 51 <= valor <= 55:
            print("MASTERCARD")
        elif lista[0] == 4:
            print("VISA")
        else:
            print("INVALID")
        
    elif le == 13:
        if lista[0] == 4:
            print("VISA")
        else:
            print("INVALID")
    
    else:
        print("INVALID")
else:
    print("INVALID")
